Json api

will-moore opened the pull request:
Documentation for the new json API.

cc @jburel

will-moore committed:
Initial commit of json-api.txt

will-moore committed:
Scripts style guide refers to 'web and insight' clients

will-moore committed:
Add 'Datasets' and improve formatting

hflynn commented on omero/developers/json-api.txt:
````diff
@@ -0,0 +1,460 @@
+        "offset": 0
+    },
+
+Sysadmins can configure the default ```limit``` and ```maxLimit``` settings
````
Did you intend to have 3 ` here?

hflynn commented on omero/developers/json-api.txt:
````diff
@@ -0,0 +1,460 @@
+
+Normalizing Experimenters and Groups
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+When returning a list of JSON objects that each contain ```omero:details``` with
````
All the mark up in this section has an extra set ` that is rendered in the staging pages too

hflynn commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,460 @@
+You can login to create an OMERO session. You must also include the CSRF token.
+The EventContext for this session will be returned to you.
+
+    POST    /api/v0/login/
```
This is indented more than the above examples, giving purple highlighting in plone

hflynn commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,460 @@
+Image containers
+----------------
+
+OMERO organises images in 2 types of many-to-many hierarcies:
```
two

hflynn reviewed (commented):
Some initial comments and can you also replace url(s) with URL(s) throughout and replace any '&' with 'and'

will-moore committed:
Initial fixes from PR #1627 feedback

hflynn commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,462 @@
+Image containers
+----------------
+
+OMERO organises images in two types of many-to-many hierarcies:
```
hierarchies typo

hflynn commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,462 @@
+
+Omero-marshal and Projection-based APIs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+The majority of the API urls use `omero-marshal <https://github.com/openmicroscopy/omero-marshal>`_
```
You've still got lowercase url throughout

hflynn commented:
Formatting looks good now, still a couple of issues but perhaps you'll list for technical review once you're ready @will-moore

will-moore committed:
Lots of tweaks and additions to json-api

mtbc commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,526 @@
+    <server>/api/v0/m/
+
+
+The webclient currently uses a small number of URLs to perform customised
```
customized

jburel replied:
could you fix it @will-moore?

mtbc commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,526 @@
+
+
+Unhandled exceptions are handled with a ``500`` error response with an
+error "message" and may also include a "stactrace".
```
(I assume here "stac" intentionally doesn't have the "k" I'm used to from elsewhere?)

will-moore replied:
my typo!

mtbc commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,526 @@
+Image containers
+----------------
+
+OMERO organises images in two types of many-to-many hierarchies:
```
organizes

mtbc replied:
maybe "hierarchy" as the "types of" is already plural? happy for @hflynn to decide that

mtbc commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,526 @@
+
+    normalize   Boolean     Place Experimenters and Groups into top-level lists instead
+                            of nesting within objects
+    childCount  Boolean     Use ?childCount=true to include an omero:childCount attribute
```
I wonder about trying some backticks around the literal code bits for the description in this kind of text

will-moore replied:
That's already in a code block table, so backticks won't be formatted right. I don't know if it's worth adding any other kind of quote marks here?

mtbc replied:
Ah, don't worry about it then.

mtbc commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,526 @@
+    normalize   Boolean     Place Experimenters and Groups into top-level lists instead
+                            of nesting within objects
+    childCount  Boolean     Use ?childCount=true to include an omero:childCount attribute
+                            for container objects
```
not quite sure how we're deciding here which descriptions get a terminal period and which don't

will-moore replied:
I'll remove them all

mtbc commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,526 @@
+
+    image       Number      Filter Datasets by child Image ID
+
+    orphaned    Boolean     Find Dataset that are not in any Project
```
Datasets

mtbc commented on omero/developers/scripts/style-guide.txt:
```diff
@@ -21,12 +21,12 @@ Most of the points below are implemented in the example :source:`Edit_Descriptio
-   will replace underscores with spaces in the script selection menu.
--  File paths - OMERO.insight will use the parent folder to build a
+-  Script Name should be in the form 'Script\_Name.py'. The OMERO.web and Insight
+   clints will replace underscores with spaces in the script selection menu.
```
clients

dominikl commented:
Looks good from my point of view, clear and understandable.

will-moore committed:
Added lots more to json-api.txt

will-moore committed:
More info text and list Wells by Index etc

will-moore committed:
Fix typo Paramters

will-moore committed:
Adding Saving and Deleting section

manics commented:
I think this is clear from the perspective of an experienced OMERO developer who understands HQL and the basics of the OMERO model. If it's aimed at developers who are attracted to the web API as their first experience of OMERO I think it'd be difficult.

Are the responses documented anywhere, for example could they be linked to omero-marshall or omero-model object docs that explain each field?

Is it worth a paragraph that explicitly states user/password/csrf/session/cookie login is used in contrast to other web APIs that a web-dev might be familiar with?

dominikl commented:
Started to implement a basic Java client example using the JSON API. Where is good place/repository to put it? Difficult to add it to openmicroscopy/examples/Training/java/src/training/ as it needs additional dependencies ( https://jsonp.java.net/ and https://hc.apache.org/index.html )

will-moore committed:
Updates from feedback in PR #5190

will-moore commented:
@manics I'm assuming that users of the API will at least have used the clients to browse data before, but they may have no idea about the model, hierarchies etc. I'll look at adding helpful links for more info on the model. I don't expect anyone to understand HQL - hopefully this API is more user-friendly than that!

Every object has a link to the Schema under ```@type```, e.g. http://www.openmicroscopy.org/Schemas/OME/2016-06#Project
Although this doesn't link to the nice interactive browser (or even link to the Project info anywhere). I think we could make that landing page more user-friendly

I'm not sure it's useful to compare our login with any other web API logins. We don't know what users will be familiar with.

will-moore committed:
Use Title case for Projects, Datasets etc

dominikl commented:
Slightly offtopic, but I wonder if would be worth making the Java Pojos "JSON compatible", so that they could be de/serialized into/from JSON objects using libraries like https://github.com/google/gson ? Well, that could also make the Java Gateway superfluous... @jburel

mtbc commented:
It's too many years ago now for me to remember details but when I was writing Java server code that talked JSON over a RESTful web API, I vaguely recall that Spring offered various convenience stuff for mapping easily between Java beans and JSON / XML representations, we just had to add some annotations and whatnot; the mapping between the transactional business methods (arguments, return value) and the JSON-over-HTTP was thus pretty painless. As we already have (equivalently old!) Spring dependency we may have some useful helpers available as part of Spring. GSON may now also be a fine approach though that wasn't an available alternative back at that time.

jburel commented on omero/developers/scripts/style-guide.txt:
```diff
@@ -120,7 +120,7 @@ Script outputs
 
 -  Scripts that generate a File Annotation or Original File should
-   return these objects. OMERO.insight will give users the option of
+   return these objects. OMERO.web and Insight will give users the option of
```
maybe use ``The clients`` here too

jburel commented on omero/developers/scripts/style-guide.txt:
```diff
@@ -120,7 +120,7 @@ Script outputs
+   return these objects. OMERO.web and Insight will give users the option of
    downloading the File, and may also allow viewing of the file if it is
    of a suitable type. This should be set as the mimetype of the File
    Annotation (E.g. 'plain/text', 'image/jpeg' etc). In this example,
```
typo ``e.g. and etc.``

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1321 @@
+
+Loading of linked objects
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
```
the extra line can be removed

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1321 @@
+
+List supported versions
+^^^^^^^^^^^^^^^^^^^^^^^
+
```
extra line

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1321 @@
+
+List starting URLs
+^^^^^^^^^^^^^^^^^^
+
```
same

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1321 @@
+
+List available OMERO servers
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
```
same

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1321 @@
+
+Get CSRF token
+^^^^^^^^^^^^^^
+
```
same

jburel reviewed (commented):
few more comments
@mtbc comments should also be addressed

jburel commented:
Add extra line under subsection with  ``^^^^^^``

will-moore committed:
Tidy and reorganise BlitzGateway documentation page

will-moore committed:
Various fixes for PR #1627

dominikl committed:
Add java example

will-moore committed:
Merge pull request #3 from dominikl/add_java_example

Add java example

will-moore committed:
Updated OMERO.web landing page

will-moore committed:
Merge remote-tracking branch 'origin/develop' into BlitzGateway_5.3.0

will-moore committed:
Add OMERO.web and Python changes to whatsnew.txt

will-moore committed:
Remove 'framework' and 'as' as suggested in PR #1661

will-moore committed:
travis fix

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1329 @@
+Error handling
+^^^^^^^^^^^^^^
+
+Invalid parameters or invalid JSON will result in a ``400 Bad Request``.
```
might be clearer with a bullet list

will-moore replied:
Sorry - not sure I understand. Bullet list of what?

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1329 @@
+    server                Number      ID of the server
+    username              String      User's username
+    password              String      User's password
+    csrfmiddlewaretoken   String      CSRF token (can be provied in header)
```
typo ``provided``

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1329 @@
+Projects but does not also load their child Datasets.
+However, it is sometimes useful to load a number of closely related objects.
+For example, loading Images also loads their Pixels data (but not Channels)
+and loading Wells also loads WellSamples and Images (but not Pixels).
```
The term "fields" is used later on in the page.
So might be good to put WellSamples (fields) here too

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1329 @@
+
+**Response**
+
+The response for a single Plate includes information on the Well Samples (fields) for
```
WellSamples

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1329 @@
+List Plate Acquisitions
+^^^^^^^^^^^^^^^^^^^^^^^
+
+A Plate Acquisition is a collection of WellSamples, grouped by an acquisition time.
```
Use "run" in brackets since it is the term used in the clients

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1329 @@
+
+The JSON API currently supports creating and saving of a limited number of
+object types, namely Projects, Datasets and Screens.
+This is due to technical issues with saving unloaded objects, such as an Image
```
I will rephrase it so it does not sound that bad
"It is not yet possible to save objects with unloaded objects e.g. ..."

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1329 @@
+without Pixels or Channels loaded. We will be working to resolve these issues
+in future releases.
+
+Creating and saving of JSON objects is handled by a single ``save`` URL
```
should be ``are handled``

will-moore committed:
Adding ?maps and histograms to WebGateway.txt

will-moore committed:
Fix WARNING on whatsnew line 16

will-moore committed:
Add note about app config to CreateApp.txt

will-moore committed:
Fix line 16 again

will-moore committed:
Fix it's -> its

will-moore committed:
Fix link in whatsnew.txt

will-moore committed:
Remove changes to whatsnew.txt

will-moore committed:
Changes to whatsnew.txt

will-moore committed:
Merge remote-tracking branch 'origin/develop' into json_api

will-moore committed:
Merge branch 'BlitzGateway_5.3.0' into json_api

will-moore commented:
From PR #1661 (merged into this one):
Added to What's New https://www.openmicroscopy.org/site/support/omero5.3-staging/developers/whatsnew.html

Also updated https://www.openmicroscopy.org/site/support/omero5.3-staging/developers/Web.html
and https://www.openmicroscopy.org/site/support/omero5.3-staging/developers/Web/WebGateway.html

Added note about App Config to https://www.openmicroscopy.org/site/support/omero5.3-staging/developers/Web/CreateApp.html

will-moore committed:
Fixes for json_api PR #1627

will-moore committed:
Add link to SPW model docs page

jburel commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -1,19 +1,69 @@
+with the OMERO API, handling connection to the server, loading of data objects
+and providing convenience methods to access the data. It was originally
+designed as part of the |OmeroWeb|, to provide connection and data
+retrieval services to various web clients. However, we have now decided to
```
I will replace ``we have now decided to`` simply by ``we encourage ...``

jburel commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -1,19 +1,69 @@
 ::
 
+    import omero
+    from omero.model import *
```
better to avoid ``import *`` and be more explicit

jburel commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -1,19 +1,69 @@
+    print p.getName().getValue()                # getValue() to unwrap the rtype
+    print p.name.val                            # short-hand
+
+    from omero.gateway import *
```
same

jburel commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -1,19 +1,69 @@
+
+::
+
+    >>> from omero.gateway import *
```
same

jburel commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -83,6 +145,19 @@ have on the client-side will become stale. We will attempt to document
+::
+
+    image = conn.getObject("Image", image_id)
+    # Initialises the Rendering engine and sets rendering settings
```
``Initializes``

jburel commented on omero/developers/Web.txt:
```diff
@@ -7,67 +7,74 @@ OMERO.web framework
-:doc:`/developers/Web/WritingViews` for more details)
+The OMERO.web framework is all based on the OMERO Python API. Code-generated
+omero.model objects communicate remotely with their counterparts on the
+OMERO.server using ICE.
```
add link to zeroC

jburel commented on omero/developers/Web.txt:
```diff
@@ -7,67 +7,74 @@ OMERO.web framework
+   More information available under :ref:`OMERO.web <omero-web>`.
+
+-  **webgateway:** Provides rendered images and JSON data for other applications,
+   or as an external API.
```
The ``or as an external API`` is not really clear

jburel commented on omero/developers/Web.txt:
```diff
@@ -7,67 +7,74 @@ OMERO.web framework
 
-Web apps
-^^^^^^^^
+-  **omero-figure:** The `OMERO.figure <http://figure.openmicroscopy.org/>`_ app
```
link to the gh repo to match the other app?

jburel commented on omero/developers/Web.txt:
```diff
@@ -7,67 +7,74 @@ OMERO.web framework
-        Main web client for viewing images, annotating, administration
-        of user and group settings etc. More information available
-        under :ref:`OMERO.web <omero-web>`.
+-  **omero-iviewer:** The `OMERO.iviewer <https://github.com/ome/omero-iviewer>`_
```
not release yet

jburel commented on omero/developers/Web.txt:
```diff
@@ -7,67 +7,74 @@ OMERO.web framework
+-  **webtagging:** The `webtagging <https://github.com/MicronOxford/webtagging>`_ app
+   was developed externally by Douglas Russell. It supports 'auto' tagging based on
+   image name and Tag-based filtering of data.
 
```
Add mapr, gallery  to the list too

jburel commented on omero/developers/Web/WebGateway.txt:
```diff
@@ -206,6 +218,20 @@ JSON methods
+       {"data": [ 24354, 93878, 87555, 45323, 27365, 14690, 9346, 2053, 60, 7, 19, 14, 15, 9, 5, 5, 3, 0, 2, 1]}
+
 
 Saving etc
```
etc.

jburel commented:
New diagram: will be nice to have other apps e.g.mapr/iviewer and not only figure

will-moore committed:
Fixes from suggestions

will-moore committed:
Use bulleted list for JSON API errors

will-moore commented:
I think it will be tricky to add more apps to the diagram without making it too confusing/cluttered. There is limited space and it's supposed to give an outline of the stack, not list everything we do.

mtbc commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -83,6 +145,19 @@ have on the client-side will become stale. We will attempt to document
+
+    # Can continue to use the connection until done,
+    # then close ALL services:
+    conn.close()
```
would it be good to have our closing example use `finally:`? (just an idea)

mtbc commented on omero/developers/Web.txt:
```diff
@@ -7,67 +7,81 @@ OMERO.web framework
-:doc:`/developers/Web/WritingViews` for more details)
+The OMERO.web framework is all based on the OMERO Python API. Code-generated
+omero.model objects communicate remotely with their counterparts on the
+OMERO.server using `ICE, from ZeroC <https://zeroc.com/products/ice>`_.
```
they capitalize it as "Ice" these days

hflynn commented:
@will-moore maybe just add 'e.g.' to the text on the figure box then, to make it clearer it's just an example

mtbc commented on omero/developers/Web.txt:
```diff
@@ -7,67 +7,81 @@ OMERO.web framework
+-  **omero-gallery:** `OMERO.gallery <https://github.com/ome/omero-gallery/>`_
+   provides a simple interface for browsing Projects, Datasets and Images.
 
 .. note::
```
I wonder if we should bump this up to being a warning.

hflynn replied:
I think so, yes

mtbc commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The majority of the API URLs use `omero-marshal <https://github.com/openmicroscopy/omero-marshal>`_
+to generate json dictionaries from OMERO **model** objects.
```
capitalization of JSON seems a bit random through the text

mtbc commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+of these objects will not be affected by the ``?normalize=true`` parameter.
+
+
+Child Counts
```
small "counts" perhaps

mtbc commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+
+Most data in OMERO has an ``Owner`` and is assigned to a permissions
+``Group``. By default, queries will return data from all owners
+accross all groups that are accessible to the current user.
```
across

mtbc commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+Projects, Datasets and Images
+-----------------------------
+
+OMERO organises Images in two types of many-to-many hierarchies:
```
organizes

mtbc replied:
probably "hierarchy" singular (like "four types of duck")

mtbc commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+    GET   /api/v0/m/wells/{well_id}/
+
+
+Creating and Saving Objects
```
should this title and those below have no capitals beyond the first letter?

mtbc commented on omero/examples/java/JSONClient.java:
```diff
@@ -0,0 +1,484 @@
+
+        for (int i = 0; i < args.length; i++) {
+            try {
+                if (args[i].trim().startsWith("--omero.webhost")) {
```
can we `trim()` it just once then use `else if`?

will-moore replied:
@dominikl Do you want to update this?

mtbc replied:
I'm happy for this nitpick not to block this PR or to be handled in 5.3.1 if preferred.

dominikl replied:
I'll fix it later when the JSONClient example gets moved to `training/examples`, where it'll be run by the CI training job.

will-moore committed:
Update OmeroWeb.png and fix typos etc

will-moore committed:
Use try/finally for conn.close() example

hflynn commented:
Is this good to merge now then @mtbc ?

mtbc commented:
My issues all look to be fixed: no remaining blockers from me!

jburel commented:
Diagram has been updated then @will-moore 
I will go over the page this evening

jburel commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -1,19 +1,69 @@
+(see :javadoc:`full list <slice2html/omero/model.html>`)
+are code-generated and mapped to the OMERO database schema. They are
+language agnostic and their data is in the form of omero.rtypes as
+described in :ref:`about model objects <AdvancedClientDevelopment#Objects>`).
```
extra ``)``

jburel commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -29,6 +79,18 @@ to handle object loading and hierarchy traversal. For example:
+As more functionality is provided by the Blitz Gateway, the coverage of object
+wrappers will increase accordingly.
+
 Access to the OMERO API services
```
Under  ``Access to the OMERO API services``

not cover by the changes but will require some adjustments
``
Stateless services (those retrieved with get… methods e.g. getQueryService) are always reused for each call, e.g. blitzon.getQueryService() whereas stateful services e.g. createRenderingEngine may be created each time.
``
I will replace ``getXXX`` also ``blitzon.getQueryService()`` does not seem right

jburel commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -83,6 +145,21 @@ have on the client-side will become stale. We will attempt to document
+
 Overwriting and extending omero.gateway classes
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
```
under  ``Overwriting and extending omero.gateway classes``
Strange to start to have Note just after NB since that what it means i.e.
``NB: Note the call to omero.gateway.refreshWrappers() to ensure that your subclasses are returned by calls to getObjects() For example:``
That sentence does not make much sense

jburel commented on omero/developers/Web.txt:
```diff
@@ -7,67 +7,81 @@ OMERO.web framework
-:doc:`/developers/Web/WritingViews` for more details)
+The OMERO.web framework is all based on the OMERO Python API. Code-generated
+omero.model objects communicate remotely with their counterparts on the
+OMERO.server using `Ice, from ZeroC <https://zeroc.com/products/ice>`_.
```
``Ice, from ZeroC``: ``,`` not needed

jburel commented on omero/developers/Web.txt:
```diff
@@ -94,8 +108,8 @@ repository.
+`http://`\ `<servername>/webtest/dataset/`\ `<datasetId>`.
 
 -  **url** goes in omeroweb/omero_webtest/urls.py This maps the URL
    'webtest/dataset/<datasetId>/' to the View function
```
code snippet needs to be updated:
* url
``url(r'^dataset/(?P<dataset_id>[0-9]+)/$', views.dataset,
        name="webtest_dataset"),``
* view
```
def dataset(request, dataset_id, conn=None, **kwargs):
    """ 'Hello World' example from tutorial on\
    http://trac.openmicroscopy.org.uk/ome/wiki/OmeroWeb """
    # before OMERO 4.3 this was conn.getDataset(dataset_id)
    ds = conn.getObject("Dataset", dataset_id)
    # generate html from template
    return render(request, 'webtest/dataset.html', {'dataset': ds})
```

will-moore replied:
I think I can remove/omit comments about OMERO 4.3 or 4.4 since I don't think many users now will be upgrading from 4.x.

hflynn replied:
Yes, I've removed 4.4 references elsewhere now, it's been unsupported for a long time

jburel commented on omero/developers/Web/WebGateway.txt:
```diff
@@ -168,6 +168,18 @@ Individual parameters are:
        ?m=g    # greyscale (only the first active channel will be shown in grey)
        ?m=c    # colour
 
+-  **Codomain maps**: OMERO's rendering engine supports mapping from input -> output
```
4 channel image => 4 channels image
also several "colours" kicking around before codomain maps

jburel commented on omero/developers/Web/WebGateway.txt:
```diff
@@ -168,6 +168,18 @@ Individual parameters are:
+   intensity map is supported, but the use of JSON encoding for the ``maps`` query
+   parameter is designed to support more maps in future.
+   In the case of the 'reverse' map, we only need to specify whether it is enabled
+   for each channel. For a 2 channel image, to enable ``reverse`` map for the
```
2 channels image

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+within each ``omero:details`` and place them under top-level ``experimenters``
+and ``experimenterGroups`` lists.
+You can specify this with the ``?normalize=true`` query parameter.
+NB: Currently this normalizing will only apply to the top-level objects
```
Minor should be ``N.B.``

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+JSON content with a ``message`` to provide more information:
+
+ -  **404 Not Found**: Caused by an invalid URL or when a specified
+    object can not be found in OMERO.
```
cannot

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+
+        {"message": "invalid literal for int() with base 10: 'foo'"}
+
+ -  **405 Method Not Allowed**: This will be returned if you try to use the
```
Replace ``This will be returned`` by ``Returned if ...`` to match the other item

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+It uses the python ``requests`` library to connect to the JSON api, login, query data, create and delete Projects.
+These steps are covered in more detail below.
+
+For an example how to use the API with Java, see `JSONClient.java <https://github.com/openmicroscopy/ome-documentation/blob/develop/omero/examples/java/JSONClient.java>`_.
```
Use "internal doc link" instead of https

will-moore replied:
I tried a bunch of things here. This works (to include the fiel) ``` .. literalinclude:: ../examples/java/JSONClient.java ``` but using ```doc``` or ```file``` doesn't work for simply creating a link. Not sure what I'm missing here?

hflynn replied:
I think a literalinclude should work here but maybe it doesn't for java files... You still don't need to put all this link in like this though, you can use a `:source:` link:
`'source' : (omero_github_root + 'blob/'+ branch + '/%s', ''),` (from the conf.py)

will-moore replied:
@hflynn This file is not under ```omero_github_root``` but I've used ```:omedocs:``` which I think works (although the link will be 404 until this is merged), right?

hflynn replied:
Sorry, that's what I get for looking at stuff in a rush! Yes, `:omedocs:` looks like it should work once the file is merged on develop.

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+        {
+          "Name": "New data",
+          "Description": "Example Project",
+          "url:project": "http://localhost:4080/api/v0/m/projects/11601/",
```
replace by <server> to match the rest

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+          "Name": "New data",
+          "Description": "Example Project",
+          "url:project": "http://localhost:4080/api/v0/m/projects/11601/",
+          "url:datasets": "http://localhost:4080/api/v0/m/projects/11601/datasets/",
```
same

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+        "@id": 3872,
+        "Name": "RNAi experiments",
+        "Description": "Knockout assays",
+        "url:datasets": "http://localhost:4080/api/v0/m/projects/3872/datasets/",
```
same

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+        {
+          "Name": "Test data",
+          "Description": "This is the Dataset description",
+          "url:dataset": "http://localhost:4080/api/v0/m/dataset/112/",
```
same
few more localhost

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+        "@id": 9702,
+        "Name": "My data",
+        "Description": "An example set",
+        "url:images": "http://localhost:4080/api/v0/m/datasets/9702/images/",
```
same

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+          "omero:details": {
+            # omitted for brevity
+          },
+          "url:image": "http://localhost:4080/api/v0/m/images/16783/",
```
same

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+          "@id": 582,
+          "Name": "Test data",
+          "Description": "This is the Screen description",
+          "url:screen": "http://localhost:4080/api/v0/m/screen/582/",
```
few more

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+          "RowNamingConvention": "letter",
+          "ColumnNamingConvention": "number",
+          "ExternalIdentifier": "003857",
+          "url:plate": "http://localhost:4080/api/v0/m/plates/5067/",
```
few more

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+        "RowNamingConvention": "letter",
+        "ColumnNamingConvention": "number",
+        "ExternalIdentifier": "003857",
+        "url:plate": "http://localhost:4080/api/v0/m/plates/5067/",
```
same

jburel commented on omero/developers/json-api.txt:
```diff
@@ -0,0 +1,1343 @@
+        {
+          "@id": 4217,
+          "url:wellsampleindex_wells": [
+            "http://localhost:4080/api/v0/m/plateacquisitions/4217/wellsampleindex/0/wells/"
```
few more

jburel commented on omero/examples/java/JSONClient.java:
```diff
@@ -0,0 +1,484 @@
+                httpGet = new HttpGet(builder.build());
+            }
+            HttpResponse res = httpClient.execute(httpGet);
+            reader = Json.createReader(new BufferedReader(
```
will the stream be closed by invoking ``JsonReader.close``?
@dominikl

dominikl replied:
It's closed in the finally block below. Ideally I should have used a try-with-resources like I did in the `put` and `post` methods. Should I correct this, which probably postpones the merge of this PR for another day or fix it later?

jburel replied:
my question was:will the inputStream be closed undirectly?

dominikl replied:
I fixed the remarked issues in https://github.com/will-moore/ome-documentation/pull/4  Leave it to @will-moore to merge the changes in.

will-moore committed:
More fixes for PR #1627

dominikl committed:
Corrected some issues with example JSONClient

jburel commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -168,9 +168,9 @@ When working with
-one. NB: Note the call to ``omero.gateway.refreshWrappers()`` to ensure
-that your subclasses are returned by calls to getObjects() For example:
-
+one. NB: The call to ``omero.gateway.refreshWrappers()`` is important to update the
```
to be in synch the other page N.B.

will-moore committed:
Use :omedocs: to link to JSONClient.java

will-moore committed:
Merge pull request #4 from dominikl/fix_json_example

Corrected some issues with example JSONClient

jburel commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -45,9 +107,9 @@ ones are not created unnecessarily. Services can be accessed using the
 Factory <slice2html/omero/api/ServiceFactory.html>`
 with the Gateway handling reuse as needed. Stateless services (those
-retrieved with get… methods e.g.
+retrieved with getXXX methods e.g.
```
do we want to make ``stateless`` then ``stateful`` in bold?

jburel commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -73,7 +135,7 @@ call to server (no lazy loading)
 Stateful services, reconnection, error handling etc
```
minor: ``etc.``

jburel commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -73,7 +135,7 @@ call to server (no lazy loading)
+The Blitz gateway was designed for use in the |OmeroWeb| and it is
 not expected that stateful services will be maintained on the client for
 significant time.
 There is various error-handling functionality in the Blitz gateway that
```
shouldn't it be ``There are various error-handling functionalities``?

jburel commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -83,6 +145,21 @@ have on the client-side will become stale. We will attempt to document
 this a little better in due course, but our general advice is to create,
```
found that a bit strange, it sounds like we did not bother explaining
I will only keep ``our general advice is to create, use and close the stateful services in the shortest practicable time``

jburel commented on omero/developers/PythonBlitzGateway.txt:
```diff
@@ -1,19 +1,69 @@
+
+    >>> from omero.gateway import BlitzGateway
+
+    >>> conn = BlitzGateway("username", "password", host="localhost", port=4064)
```
maybe add try/finally with close to make sure that people remember to close

will-moore replied:
I'll add a close, but the try/finally is a bit too distracting for a very simple example of lazy loading that doesn't create any stateful services.
However, we should probably emphasise the use of try/finally conn.close() on the main Python examples page, since we don't show it at all on that page.

joshmoore replied:
OT: for me, there's always going to be a balance between what we can teach people in our examples and what they simply have to learn for themselves. But where possible, I do think it is good to instill in them where there are potentials for real harm in their code.

When I see ` resource = svc.open()` without a `resource.close()` I get nervous. But when I see `resource.close()` without a `try/finally` I _also_ get nervous. Are they equally risky? It depends.

I wonder if the introduction of a [context manager](https://docs.python.org/2/library/contextlib.html) would make this simple enough that we'd feel comfortable putting it everywhere?

```
with BlitzGateway() as conn:
   # now no longer risky stuff
```

jburel replied:
we should certainly look at improving the various sections of the doc/scripts even the tests with context manager.
(test will be a good place to start since we will success/failure answer)
Better that it was, but still some improvements to come

jburel replied:
I have created a card https://trello.com/c/eHaRc8Zj/52-investigate-usage-of-context-manager

jburel commented on omero/developers/Web.txt:
```diff
@@ -7,67 +7,81 @@ OMERO.web framework
+^^^^^^^^^^^^
+
+The Blitz Gateway wraps omero.model objects to facilitate many loading
+and update operations (see |OmeroPy|).
```
better to point to ``PythonBlitzGateway`` instead of ``Python`` page

jburel commented on omero/developers/Web.txt:
```diff
@@ -94,16 +108,16 @@ repository.
-       url( r'^dataset/(?P<datasetId>[0-9]+)/$', views.dataset ),
+       url( r'^dataset/(?P<dataset_id>[0-9]+)/$', views.dataset, name="webtest_dataset" ),
 
 -  **View** function, in omeroweb/omero_webtest/views.py. NB: @login\_required
```
Why View and Template start with Capital and not Url
I will remove capital for view and template

jburel replied:
N.B.

jburel commented on omero/developers/Web/WebGateway.txt:
```diff
@@ -136,7 +136,7 @@ Individual parameters are:
+-  **Channel color** e.g. to set the colors for channels 1 to red and 2
    to green and 3 to blue:
 
    ::
```
``Rendering levels e.g. to set the cut-in and cut-out values for a 3 Channel image.``
replace by image with 3 channels to match the rest

jburel commented on omero/developers/Web/WebGateway.txt:
```diff
@@ -206,9 +218,23 @@ JSON methods
 -  Lots of metadata for the image. See
    below: ``webgateway/imgData/<imageId>/``
 
+-  Histogram of pixel intensity data for an image plane. Channel index is zero-based.
```
in bold?

will-moore replied:
Which bit? Presumably the **zero-based** ?

jburel replied:
I was thinking Histogram to highlight the feature
zero-based could be too

jburel reviewed (commented):
few more than I missed previously

will-moore committed:
More fixes

jburel commented:
Looks good now.
Thanks for improving the pages

will-moore committed:
Bold **Histogram** instead of zero-based

hflynn commented:
All sorted barring issues that are carded for 5.3.1, merging thanks.
